File: leetcode/0001-1000/0769-max-chunks-to-make-sorted/MaxChunksToMakeSorted.ts
function maxChunksToSorted(integers: number[]): number {
  const { totalChunks } = integers.reduce(
    (previous, currentElement, currentIndex) => {
      let { totalChunks, max } = previous;

      max = Math.max(max, currentElement);

      if (currentIndex === max) {
        totalChunks++;
      }

      return {
        totalChunks,
        max
      };
    }
    , { totalChunks: 0, max: 0 }
  );

  return totalChunks
};

console.log(maxChunksToSorted([1, 0, 2, 3, 4]));

File: leetcode/0001-1000/0701-0800/0769-max-chunks-to-make-sorted/MaxChunksToMakeSorted.java
public class MaxChunksToMakeSorted {
  public static void main(String[] args) {
    System.out.println((new Solution()).maxChunksToSorted(new int[] { 0, 1, 2, 3, 4, 5 }));
  }
}

class Solution {
  public int maxChunksToSorted(int[] integers) {
    int max = 0;
    int totalChunks = 0;

    for (int i = 0; i < integers.length; i++) {
      max = Math.max(max, integers[i]);

      if (max == i) {
        totalChunks++;
      }
    }

    return totalChunks;
  }
}
